cpu/stm32/stmclk/stmclk_mp1.c

```diff
@@ -0,0 +1,230 @@
+#define CLOCK_ENABLE_HSE                1
+#else
+#define CLOCK_ENABLE_HSE                0
+#endif
+
+/* Check whether HSI must be enabled:
+  - When HSI is used as SYSCLK
+  - When PLL is used as SYSCLK and the board doesn't provide HSE (since HSI will be
+    used as PLL input clock)
+*/
+#if IS_ACTIVE(CONFIG_USE_CLOCK_HSI) || \
+    (!IS_ACTIVE(CONFIG_BOARD_HAS_HSE) && IS_ACTIVE(CLOCK_ENABLE_PLL))
+#define CLOCK_ENABLE_HSI                1
+#else
+#define CLOCK_ENABLE_HSI                0
+#endif
+
+void stmclk_enable_hsi(void)
+{
+    RCC->OCENSETR |= RCC_OCENSETR_HSION;
+    while (!(RCC->OCRDYR & RCC_OCRDYR_HSIRDY)) {}
+}
+
+void stmclk_disable_hsi(void)
+{
+    /* we only disable the HSI clock if not used as input for the PLL and if
+     * not used directly as system clock */
+    if (CLOCK_HSE) {
+        RCC->OCENCLRR |= RCC_OCENCLRR_HSION;
+        while (!(RCC->OCRDYR & RCC_OCRDYR_HSIRDY)) {}
+    }
+}
+
+static void stmclk_enable_hse(void)
+{
+    RCC->OCENSETR |= RCC_OCENSETR_HSEON;
+    while (!(RCC->OCRDYR & RCC_OCRDYR_HSERDY)) {}
+}
+
+void stmclk_init_sysclk(void)
+{
+    /* disable any interrupts. Global interrupts could be enabled if this is
+     * called from some kind of bootloader...  */
+    unsigned is = irq_disable();
+    RCC->MC_CIFR = 0;
+
+    /* enable HSI clock for the duration of initialization */
+    stmclk_enable_hsi();
+
+    /* use HSI as system clock while we do any further configuration and
+     * configure the AHB and APB clock dividers as configure by the board */
+    /* MCUDIV */
+    RCC->MCUDIVR =  CLOCK_MCU_DIV;
+    while ((RCC->MCUDIVR & RCC_MCUDIVR_MCUDIVRDY)
+            != RCC_MCUDIVR_MCUDIVRDY) {}
+    /* APB1DIV */
+    RCC->APB1DIVR =  CLOCK_APB1_DIV;
+    while ((RCC->APB1DIVR & RCC_APB1DIVR_APB1DIVRDY)
+            != RCC_APB1DIVR_APB1DIVRDY) {}
+    /* APB2DIV */
+    RCC->APB2DIVR =  CLOCK_APB2_DIV;
+    while ((RCC->APB2DIVR & RCC_APB2DIVR_APB2DIVRDY)
+            != RCC_APB2DIVR_APB2DIVRDY) {}
+    /* APB3DIV */
+    RCC->APB3DIVR =  CLOCK_APB3_DIV;
+    while ((RCC->APB3DIVR & RCC_APB3DIVR_APB3DIVRDY)
+            != RCC_APB3DIVR_APB3DIVRDY) {}
+
+    /* MCU clock source */
+    RCC->MSSCKSELR = RCC_MSSCKSELR_MCUSSRC_0; /* HSI */
+    while ((RCC->MSSCKSELR & RCC_MSSCKSELR_MCUSSRCRDY)
+            != RCC_MSSCKSELR_MCUSSRCRDY) {}
+
+    /* disable all active clocks except HSI -> resets the clk configuration */
+    RCC->OCENCLRR = ~(RCC_OCENSETR_HSION);
+
+    /* if configured, we need to enable the HSE clock now */
+#if (CLOCK_HSE)
+    stmclk_enable_hse();
+#endif
+
+    /* now we can safely configure the PLL */
+    RCC->PLL3CFGR1 = (PLL_M | PLL_N);
+    RCC->PLL3CFGR2 = (PLL_P | PLL_Q | PLL_R);
+
+    RCC->RCK3SELR |= PLL_SRC;
+    while (!(RCC->RCK3SELR & RCC_RCK3SELR_PLL3SRCRDY)) {}
+
+    /* and start the PLL */
+    RCC->PLL3CR |= (RCC_PLL3CR_DIVPEN | RCC_PLL3CR_DIVQEN
+            | RCC_PLL3CR_DIVREN | RCC_PLL3CR_PLLON);
+    while (!(RCC->PLL3CR & RCC_PLL3CR_PLL3RDY)) {}
```

Comment:
You could wrap all this in a conditional if PLL is needed:

```suggestion
    if (IS_ACTIVE(CLOCK_ENABLE_PLL)) {
        /* now we can safely configure the PLL */
        RCC->PLL3CFGR1 = (PLL_M | PLL_N);
        RCC->PLL3CFGR2 = (PLL_P | PLL_Q | PLL_R);

        RCC->RCK3SELR |= PLL_SRC;
        while (!(RCC->RCK3SELR & RCC_RCK3SELR_PLL3SRCRDY)) {}

        /* and start the PLL */
        RCC->PLL3CR |= (RCC_PLL3CR_DIVPEN | RCC_PLL3CR_DIVQEN
                | RCC_PLL3CR_DIVREN | RCC_PLL3CR_PLLON);
        while (!(RCC->PLL3CR & RCC_PLL3CR_PLL3RDY)) {}
    }
```

Reply:
For me it is a "static" test, thus a macro is more appropriate. Do you agree ?
If yes, it should be the same in other stm32: https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/stmclk/stmclk_f2f4f7.c#L535

Reply:
If we allow several clock configurations, then only using preprocessor `#if` will skip the non-default parts in the CI. Using regular C `if` and everything is compiled, even if unused, and the compiler will optimize these unused parts anyway. So it's better to not used preprocessor `#if` in this case.

Reply:
Ack thx